### PR TITLE
[DSIP-44] Set a delay time to TaskExecuteRunnable if it dispatched failed

### LINKED ISSUE #16065: [DSIP-44] Set a delay time to TaskExecuteRunnable if it dispatched failed
### Search before asking

- [X] I had searched in the [DSIP](https://github.com/apache/dolphinscheduler/issues/14102) and found no similar DSIP.

### Motivation

We will use a DelayQueue to store the waiting dispatch tasks, once the task dispatch failed, it will be put back to the queue.

The problem is that when we have a worker overload or worker group crash, the dispatch will failed until the worker is available. So we need to wait until the worker is available.

### Design Detail

The simple way is add a wait time for different task.

image

The wait time should be increase after each failed. e.g. 1s, 5s, 10s, 60s ... max_waiting_time

### Compatibility, Deprecation, and Migration Plan

Compatibility

### Test Plan

Test by UT

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

 Fix GlobalTaskDispatchWaitingQueueLooper might be busy when worker is unavaliable

close https://github.com/apache/dolphinscheduler/issues/16065

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java [hunk 1]
```diff
@@ -52,4 +52,30 @@ public TaskExecutionContext getTaskExecutionContext() {
...
+            return taskInstancePriorityCompareResult;
+        }
+        // larger number, higher priority
+        int taskGroupPriorityCompareResult =
+                taskInstance.getTaskGroupPriority() - other.getTaskInstance().getTaskGroupPriority();
+        if (taskGroupPriorityCompareResult != 0) {
+            return -taskGroupPriorityCompareResult;
+        }
+        // The task instance shouldn't be equals
+        return taskInstance.getFirstSubmitTime().compareTo(other.getTaskInstance().getFirstSubmitTime());
```
Comment (Radeity) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:80]: Maybe earlier submit time should have higher priority?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java [hunk 2]
```diff
@@ -52,4 +52,30 @@ public TaskExecutionContext getTaskExecutionContext() {
...
+        int taskInstancePriorityCompareResult = taskInstance.getTaskInstancePriority().getCode()
+                - other.getTaskInstance().getTaskInstancePriority().getCode();
+        if (taskInstancePriorityCompareResult != 0) {
+            return taskInstancePriorityCompareResult;
+        }
+        // larger number, higher priority
+        int taskGroupPriorityCompareResult =
+                taskInstance.getTaskGroupPriority() - other.getTaskInstance().getTaskGroupPriority();
+        if (taskGroupPriorityCompareResult != 0) {
+            return -taskGroupPriorityCompareResult;
```
Comment (Radeity) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: If larger number has higher priority, this place should return `taskGroupPriorityCompareResult `, and return value in line 63&69 should return opposite value.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java [hunk 3]
```diff
@@ -64,29 +58,34 @@ public synchronized void start() {
...
+                log.warn("The TaskInstance {} state is : {}, will not dispatch", taskInstance.getName(), status);
+                return;
             }
+            taskDispatchFactory.getTaskDispatcher(taskInstance).dispatchTask(taskExecuteRunnable);
+        } catch (Exception e) {
+            // If dispatch failed, will put the task back to the queue
+            // The task will be dispatched after waiting time.
+            // the waiting time will increase exponentially, but will not exceed 60 seconds
+            long waitingTimeMills = Math.max(
+                    taskExecuteRunnable.getTaskExecutionContext().increaseDispatchFailTimes() * 1_000L, 60_000L);
```
Comment (Radeity) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java:86]: It seems add 1s for each time, not exponentially.

Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/GlobalTaskDispatchWaitingQueueLooper.java:86]: The wait time will be `1,2,3,4,5...60`. I updated the comment

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: This is a history problem, the higher task group priority the small number.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:80]: Yes, the earlier submit time will have higher priority, I add ut for this case

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java (hunk 2 above)
Comment (Radeity) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: Ohh, so the comment here should be `smaller number, higher priority`?

Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: Updated the comment, for task instance priority // smaller number, higher priority, for task group priority // larger number, higher priority

**Review by Radeity — DISMISSED** [review]
LGTM

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java (hunk 2 above)
Comment (Radeity) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: Wait, two comments seem reversed, for task group priority, `smaller number, higher priority`, for workerflow/taskInstance, `larger number, higher priority`.

Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: No, for task instance/workflow instance smaller number, higher priority.
image

Comment (Radeity) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: I see, but according to UT and code logic, task instance of `Priority.HIGH` has higher priority than `Priority.LOW`

Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/BaseTaskExecuteRunnable.java:77]: Yes, Priority.HIGH have small number

**Comment by ruanwenjun** [comment]
No new change, just rebase commit.

**Review by Radeity — APPROVED** [review]

**Review by EricGao888 — APPROVED** [review]

**Pull request merged**